### ISSUE TITLE
tag-bans doesnt seem to work for script tags

### ISSUE THREAD

**Issue by TDA** [htmllint/htmllint#108]
Hey,
So we have been using htmllint to lint files, and potentially flag them for containing script tags (amongst other tags).

Here's my htmllintrc file:
{
    "attr-bans": ["src", "dynsrc", "lowsrc", "onclick"],
    "attr-no-dup": true,
    "attr-no-unsafe-char": true,
    "attr-quote-style": "quoted",
    "doctype-first": false,
    "id-class-style": false,
    "indent-style": "spaces",
    "indent-width": 0,
    "tag-bans": ["img", "script", "input", "form", "textarea"],
    "tag-name-lowercase": true,
    "tag-name-match": true
}

and here is the html file:
```
<html>
<head>
<title>just a normal title here</title>
</head>
<body>
<script src = "">console.log('hi');
</script>
<input type = "text" value="nothing">
<a src = "asd.js" dynsrc= "" onclick="">dsaf</a>
</body>
</html>
```

The output I am receiving is : 
$ htmllint xyz.html
[htmllint error in xyz.html ] TypeError: Cannot read property '0' of undefined
[htmllint] found 0 errors out of 1 files

If i remove the input tag: 
htmllint xyz.html
xyz.html: line 10, col 3, the "src" attribute is banned
xyz.html: line 10, col 18, the "dynsrc" attribute is banned
xyz.html: line 10, col 29, the "onclick" attribute is banned
[htmllint] found 3 errors out of 1 files

The linter doesnt work on the tag-bans list, and worse, it throws an error in the presence of input tags or img tags with/without them being in the bans list with a cryptic message and does not show any of the actual errors.

**Comment by pdehaan** [comment]
I think it may be due to the `tag-name-match` flag.

#### index.html:
```html
<input type="text" />
```

#### .htmllintrc:
```json
{
  "tag-bans": ["img", "script", "input", "form", "textarea"],
  "tag-name-match": true
}
```

#### Output:
```
$ htmllint index.html

[htmllint error in index.html ] TypeError: Cannot read property '0' of undefined
[htmllint] found 0 errors out of 1 files
```

But if I change `tag-name-match` to `false`, it seems to work:
```
$ cat .htmllintrc && htmllint index.html
{
  "tag-bans": ["img", "script", "input", "form", "textarea"],
  "tag-name-match": false
}

index.html: line 1, col 1, the input tag is banned

[htmllint] found 1 errors out of 1 files
```

---

```sh
$ htmllint --version # 0.0.4
```

**Comment by TDA** [comment]
@pdehaan But that was supposed to be fixed in https://github.com/htmllint/htmllint/issues/97.

**Comment by pdehaan** [comment]
@TDA: I don't know that #97 fix has been released. Looking through the code, it looks like I'm using **htmllint-cli@0.0.4** which includes **htmllint@0.2.4** which was released on [Apr 12](https://github.com/htmllint/htmllint/releases).
The https://github.com/htmllint/htmllint/issues/97#issuecomment-95981814 was made on Apr 24 and I believe refers to commit https://github.com/htmllint/htmllint/commit/210272591d648b9765a88720a748541d2d47770b which was added after **v0.2.4** was published to npm. So we'd need to wait for v0.2.5 or v0.3.0 (and then a newer version of the **htmllint-cli**) to see that change.

**Comment by pdehaan** [comment]
Actually, I think this will _almost_ work once a new version of **htmllint** and **htmllint-cli** gets deployed.

I hacked my node_modules/htmllint-cli to install **htmllint** from GitHub instead of a published version from npm (set your node_modules/htmllint-cli/package.json **htmllint** dependency to `"htmllint": "htmllint/htmllint",` and re-run `npm install` -- then I hacked my main package.json's `test` script to `"test": "htmllint index.html"` so it uses my ./node_modules/.bin/htmllint instead of my globally installed version):

```sh
$ cat .htmllintrc
{
  "attr-bans": ["src", "dynsrc", "lowsrc", "onclick"],
  "attr-no-dup": true,
  "attr-no-unsafe-char": true,
  "attr-quote-style": "quoted",
  "doctype-first": false,
  "id-class-style": false,
  "indent-style": "spaces",
  "indent-width": 0,
  "tag-bans": ["img", "script", "input", "form", "textarea"],
  "tag-name-lowercase": true,
  "tag-name-match": true
}


$ cat index.html
<html>
<head>
  <title>just a normal title here</title>
</head>
<body>
  <script src="">console.log('hi');</script>
  <input type="text" value="nothing">
  <a src="asd.js" dynsrc="" onclick="">dsaf</a>
</body>
</html>


$ htmllint index.html # using my global htmllint-cli@0.0.4 module

[htmllint error in index.html ] TypeError: Cannot read property '0' of undefined
[htmllint] found 0 errors out of 1 files


$ npm test # using my hacked version from htmllint GitHub master

> htmllint-test@1.0.0 test /Users/pdehaan/dev/tmp/del/htmllint-test
> htmllint index.html

index.html: line 7, col 3, the input tag is banned
index.html: line 8, col 5, the `src` attribute is banned
index.html: line 8, col 18, the `dynsrc` attribute is banned
index.html: line 8, col 28, the `onclick` attribute is banned

[htmllint] found 4 errors out of 1 files
npm ERR! Test failed.  See above for more details.
```

It looks like the `attr-bans` bans are working, but not sure why it isn't complaining that the `<script>` tag is banned.

**Comment by TDA** [comment]
@pdehaan That is encouraging, but we certainly cant have partial functionality for the banned tags. Either everything works, or it doesnt do any good.

**Comment by mlochbaum** [comment]
The script tag is parsed by htmlparser2 as an element of type "script" rather than "tag". So the only problem I see (besides our npm versions being behind) is that tag-bans doesn't work specifically on the script tag.
That issue is pretty easily fixed by adding `'script'` to the `tag-bans` rule's `filter` propery. I should also dig through the parser's documentation to see what other not-actually-tags exist.

**Comment by mlochbaum** [comment]
(Closed by commit 2242756e, but I forgot to write that in the commit message).

**Comment by TDA** [comment]
@mlochbaum Thank you so much!
